test/gracefulshutdown/gracefulshutdown_test.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	gatewayPodName := gatewayPods.Items[0].Name
+
+	// Create a service and an ingress
+	name, port, _ := ingress.CreateTimeoutService(ctx, t, clients)
+	_, client, _ := ingress.CreateIngressReady(ctx, t, clients, v1alpha1.IngressSpec{
+		Rules: []v1alpha1.IngressRule{{
+			Hosts:      []string{name + ".example.com"},
+			Visibility: v1alpha1.IngressVisibilityExternalIP,
+			HTTP: &v1alpha1.HTTPIngressRuleValue{
+				Paths: []v1alpha1.HTTPIngressPath{{
+					Splits: []v1alpha1.IngressBackendSplit{{
+						IngressBackend: v1alpha1.IngressBackend{
+							ServiceName:      name,
+							ServiceNamespace: test.ServingNamespace,
+							ServicePort:      intstr.FromInt(port),
+						},
+					}},
+				}},
+			},
+		}},
+	})
+
+	tests := []struct {
+		name            string
+		requestDuration time.Duration
+		wantStatusCode  int
+	}{
+		{
+			name:            fmt.Sprintf("do a request taking slightly less than the drain time: %s", drainTime),
+			requestDuration: drainTime - (3 * time.Second),
+			wantStatusCode:  http.StatusOK,
+		},
+		{
+			name:            fmt.Sprintf("do a request taking slightly more than the drain time: %s", drainTime),
+			requestDuration: drainTime + (3 * time.Second),
+			wantStatusCode:  0,
+		},
+	}
+
+	g := new(errgroup.Group)
+	var statusCodes sync.Map
+
+	// Run all requests asynchronously at the same time, and collect the results in statusCodes map
+	for i := range tests {
+		test := tests[i]
+
+		g.Go(func() error {
+			statusCode, err := sendRequest(client, name, test.requestDuration)
+			statusCodes.Store(test.name, statusCode)
+			return err
+		})
+	}
+
+	// Once requests are in-flight, delete the gateway pod:
+	// the 1 second sleep before the deletion is here to ensure the requests have been sent by the goroutines above
+	time.Sleep(1 * time.Second)
+	if err := clients.KubeClient.CoreV1().Pods(gatewayNs).Delete(ctx, gatewayPodName, metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("Failed to delete pod %s: %v", gatewayPodName, err)
+	}
+
+	// Wait until we get responses from the asynchronous requests
+	if err := g.Wait(); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, test := range tests {
+		statusCode, _ := statusCodes.Load(test.name)
+
+		assert.Equal(t, statusCode.(int), test.wantStatusCode, fmt.Sprintf("%s has failed: expected %d, got %s",
+			test.name, test.wantStatusCode, statusCode))
+	}
+}
+
+func sendRequest(client *http.Client, name string, requestTimeout time.Duration) (statusCode int, err error) {
+	reqURL := fmt.Sprintf("http://%s.example.com?initialTimeout=%d", name, requestTimeout.Milliseconds())
+	req, err := http.NewRequest("GET", reqURL, nil)
+	if err != nil {
+		return 0, fmt.Errorf("error making GET request: %w", err)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		var errURL *url.Error
+		// When the gateway cuts the connection (after completing the drain process), an EOF error is returned to the client
```

Comment:
Note for me: I'm really not sure about this behavior, I need to test in real conditions. I would have expected a 5xx error with a `Connection: close` header.